### PR TITLE
Prevent inconsistent state due to concurrent sorting

### LINKED ISSUE #28861: Multiple istiod instances have different state
**Bug description**

We got outselves into a state where different istiod instances have different state. When running `istioctl proxy-status` we get something this:

Name                          | CDS        | LDS                        | EDS    | RDS    | ISTIOD           | VERSION |
------------------------------|------------|----------------------------|--------|--------|------------------|---------|
&lt;pod&gt;.&lt;namespace&gt; | SYNCED     | SYNCED                     | SYNCED | SYNCED | &lt;istiod 1&gt; | 1.7.3   |
&lt;pod&gt;.&lt;namespace&gt; | SYNCED     | SYNCED                     | SYNCED | SYNCED | &lt;istiod 1&gt; | 1.7.3   |
&lt;pod&gt;.&lt;namespace&gt; | SYNCED     | STALE (Never Acknowledged) | SYNCED | SYNCED | &lt;istiod 2&gt; | 1.7.3   |
&lt;pod&gt;.&lt;namespace&gt; | SYNCED     | STALE (Never Acknowledged) | SYNCED | SYNCED | &lt;istiod 2&gt; | 1.7.3   |
&lt;pod&gt;.&lt;namespace&gt; | SYNCED     | SYNCED                     | SYNCED | SYNCED | &lt;istiod 3&gt; | 1.7.3   |
&lt;pod&gt;.&lt;namespace&gt; | SYNCED     | SYNCED                     | SYNCED | SYNCED | &lt;istiod 3&gt; | 1.7.3   |

The log of the misbehaving istiod shows this line:

```
[...] Internal:Error adding/updating listener(s) 0.0.0.0_443: error adding listener '0.0.0.0:443': multiple filter chains with overlapping matching rules are defined
```

How can this happen? `istioctl analyze` does not indicate any problems and we are unsure about what is causing the conflict. See steps to reproduce for what we _suspect_.

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

Either all istiod instance fail with the same error messages or not at all.

**Steps to reproduce the bug**

We were unable to reliably reproduce the bug. However, before this bug occurred we tried to setup two egress gateways and split traffic between them:

Apply the following Istio operator configuration to install Istio with two egress gateways:

```yaml
kubectl apply -f - <<EOF
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio
  namespace: istio-system
spec:
  #meshConfig:
    #accessLogFile: /dev/stdout
    #enableAutoMtls: false
    #outboundTrafficPolicy:
    #  mode: REGISTRY_ONLY

  components:
    egressGateways:
      - name: istio-egressgateway-default
        enabled: true

        # Patch broken default selector.
        k8s:
          overlays:
            - kind: Service
              name: istio-egressgateway-default
              patches:
                - path: spec.selector
                  value:
                    service.istio.io/canonical-name: istio-egressgateway-default

      - name: istio-egressgateway-special
        enabled: true

        # Patch broken default selector.
        k8s:
          overlays:
            - kind: Service
              name: istio-egressgateway-special
              patches:
                - path: spec.selector
                  value:
                    service.istio.io/canonical-name: istio-egressgateway-special
EOF
```

Then make sure to install the SNI proxies. We deviate slightly from what is described in
[Wildcard configuration for arbitrary domains](https://istio.io/latest/docs/tasks/traffic-management/egress/wildcard-egress-hosts/#wildcard-configuration-for-arbitrary-domains)
by deploying the SNI proxy as a separate deployment. Also, we deploy two SNI proxies: one for the "default" egress
gateway and one for the "special" egress gateway.

```yaml
kubectl apply -f - <<EOF
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: sni-proxy
  namespace: istio-system
data:
  nginx.conf: |
    events {
    }

    http {
      log_format log_http '\$remote_addr [\$time_local] HTTP [\$http_host] \$status';

      access_log /var/log/nginx/access.log log_http;
      error_log  /var/log/nginx/error.log;

      # HTTP forward proxy
      server {
        listen  0.0.0.0:8080;

        location / {
          resolver 8.8.8.8 ipv6=off;
          proxy_pass http://\$http_host\$uri\$is_args\$args ;
        }
      }
    }

    stream {
      log_format log_stream '\$remote_addr [\$time_local] \$protocol [\$ssl_preread_server_name] \$status';

      access_log /var/log/nginx/access.log log_stream;
      error_log  /var/log/nginx/error.log;

      # TCP forward proxy using SNI
      server {
        resolver 8.8.8.8 ipv6=off;
        listen       0.0.0.0:8443;
        proxy_pass   \$ssl_preread_server_name:443;
        ssl_preread  on;
      }
    }
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: sni-proxy-special
  namespace: istio-system
spec:
  replicas: 1
  selector:
    matchLabels:
      proxy: "special"
  template:
    metadata:
      labels:
        proxy: "special"
    spec:
      containers:
        - name: sni-proxy
          image: nginx:1.19.4
          ports:
            - name: http
              containerPort: 8080
              protocol: TCP
            - name: https
              containerPort: 8443
              protocol: TCP
          volumeMounts:
            - name: cache
              mountPath: /var/cache/nginx
            - name: config
              mountPath: /etc/nginx
              readOnly: true
            - name: pid
              mountPath: /var/run
              subPath: nginx.pid
      volumes:
        - name: cache
          emptyDir: {}
        - name: config
          configMap:
            name: sni-proxy
        - name: pid
          emptyDir: {}
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: sni-proxy-default
  namespace: istio-system
spec:
  replicas: 1
  selector:
    matchLabels:
      proxy: "default"
  template:
    metadata:
      labels:
        proxy: "default"
    spec:
      containers:
        - name: sni-proxy
          image: nginx:1.19.4
          ports:
            - name: http
              containerPort: 8080
              protocol: TCP
            - name: https
              containerPort: 8443
              protocol: TCP
          volumeMounts:
            - name: cache
              mountPath: /var/cache/nginx
            - name: config
              mountPath: /etc/nginx
              readOnly: true
            - name: pid
              mountPath: /var/run
              subPath: nginx.pid
      volumes:
        - name: cache
          emptyDir: {}
        - name: config
          configMap:
            name: sni-proxy
        - name: pid
          emptyDir: {}
---
apiVersion: v1
kind: Service
metadata:
  name: proxy-special
  namespace: istio-system
spec:
  ports:
    - port: 9080
      targetPort: http
      protocol: TCP
      name: http
    - port: 9443
      targetPort: https
      protocol: TCP
      name: https
  selector:
    proxy: "special"
---
apiVersion: v1
kind: Service
metadata:
  name: proxy-default
  namespace: istio-system
spec:
  ports:
    - port: 9080
      targetPort: http
      protocol: TCP
      name: http
    - port: 9443
      targetPort: https
      protocol: TCP
      name: https
  selector:
    proxy: "default"
EOF
```

To wire everything up create the virtual service, the gateways, the destiantion rules and the service entry required:

```yaml
kubectl apply -f - <<EOF
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: egress
  namespace: istio-system
spec:
  hosts:
    - "*.com"
    - "*.org"
    - "*.de"
  gateways:
    - mesh
    - gw-special
    - gw-default
  http:
    - match:
        - gateways:
            - mesh
          port: 80
      route:
        - destination:
            host: istio-egressgateway-default.istio-system.svc.cluster.local
            subset: all-default
            port:
              number: 80
          weight: 100
    - match:
        - gateways:
            - gw-default
          port: 80
      route:
        - destination:
            host: proxy-default.istio-system.svc.cluster.local
            port:
              number: 9080
          weight: 100
  tls:
    - match:
        - gateways:
            - mesh
          port: 443
          sniHosts:
            - "www.example.com"
            - "www.google.com"
            - "en.wikipedia.org"
      route:
        - destination:
            host: istio-egressgateway-special.istio-system.svc.cluster.local
            subset: all-special
            port:
              number: 443
    - match:
        - gateways:
            - mesh
          port: 443
          sniHosts:
            - "*.com"
            - "*.org"
            - "*.de"
      route:
        - destination:
            host: istio-egressgateway-default.istio-system.svc.cluster.local
            subset: all-default
            port:
              number: 443
    - match:
        - gateways:
            - gw-default
          port: 443
          sniHosts:
            - "*.com"
            - "*.org"
            - "*.de"
      route:
        - destination:
            host: proxy-default.istio-system.svc.cluster.local
            port:
              number: 9443
          weight: 100
    - match:
        - gateways:
            - gw-special
          port: 443
          sniHosts:
            - "www.google.com"
            - "www.example.com"
            - "en.wikipedia.org"
      route:
        - destination:
            host: proxy-special.istio-system.svc.cluster.local
            port:
              number: 9443
          weight: 100
---
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: gw-default
  namespace: istio-system
spec:
  selector:
    service.istio.io/canonical-name: istio-egressgateway-default
  servers:
    - port:
        number: 80
        name: http-default
        protocol: HTTP
      hosts:
        - "*"
    - port:
        number: 443
        name: tls-default
        protocol: TLS
      hosts:
        - "*"
      tls:
        mode: PASSTHROUGH
---
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: gw-special
  namespace: istio-system
spec:
  selector:
    service.istio.io/canonical-name: istio-egressgateway-special
  servers:
    - port:
        number: 443
        name: tls-special
        protocol: TLS
      hosts:
        - "www.example.com"
        - "www.google.com"
      tls:
        mode: PASSTHROUGH
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: any
spec:
  hosts:
    - "*.com"
  ports:
    - number: 443
      name: https
      protocol: HTTPS
---
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: dest-special
  namespace: istio-system
spec:
  host: istio-egressgateway-special.istio-system.svc.cluster.local
  subsets:
    - name: all-special
---
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: dest-default
  namespace: istio-system
spec:
  host: istio-egressgateway-default.istio-system.svc.cluster.local
  subsets:
    - name: all-default
EOF
```

To verify everything works, deploy the sleep pod:

```
curl https://raw.githubusercontent.com/istio/istio/master/samples/sleep/sleep.yaml | kubectl apply -f -
```

And execute a few `curl` commands on it. You will see the requests showing up in the SNI proxies.

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**

```
$ istioctl version --remote
client version: 1.7.4
control plane version: 1.7.3
data plane version: 1.7.3 (49 proxies)
```

```
$ kubectl version --short
Client Version: v1.19.4
Server Version: v1.17.9-eks-4c6976
```

**How was Istio installed?**

Istio was installed using the Istio operator.

**Environment where bug was observed (cloud vendor, OS, etc)**

We are running an EKS cluster on AWS.

We are running quite a few workloads. Providing a full cluster dump would add a lot of noise. Happy to provide dumps of likely suspects though!

### PR DESCRIPTION
**Pull request by jnodorp**
Fixes #28861

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by jnodorp** [comment]
@howardjohn this would be a quick fix to #28861. However, it does not solve the concurrent access.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 1]
```diff
@@ -1922,12 +1922,12 @@ func buildListener(opts buildListenerOpts) *listener.Listener {
 				}
 			}
 			if !fullWildcardFound {
-				sort.Strings(chain.sniHosts)
+				sort.Stable(sort.StringSlice(chain.sniHosts))
```
Comment (jnodorp) [inline pilot/pkg/networking/core/v1alpha3/listener.go:1925]: We haven't discussed this yet, but this might be affected as well.

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1925]: I agree stable sort will fix this case but it still seems like we have a fundamental race here which may lead to future bugs down the line (or already does!). Would it make more sense to just copy these slices?

Comment (jnodorp) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1925]: The question remaining is: where to copy the slice? When copying right here we might be copying the "wrong" state already. There are various places where SNI hosts and destination CIDRs are written. The ones I peeked at didn't seem particularly suspicious, but I'm not familiar with Istios code.

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1925]: can we do the sorting in `initVirtualServices` so this is once for all and no need copy

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1925]: Is the sni hosts list always copied directly from virtual service without adding or removing any items?

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/listener.go:1925]: Also from gateway, but we have sorted it there.  `GetSNIHostsForServer`

**Pull request closed without merging**